<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="panda">

    <xacro:arg name="fixed_base" default="false"/>

    <xacro:include filename="$(find mppi_panda)/resources/panda/panda.urdf.xacro"/>
    <xacro:include filename="$(find mppi_manipulation)/data/planar_base.urdf.xacro"/>
    <xacro:include filename="$(find mppi_manipulation)/data/ridgeback.urdf.xacro"/>

    <!--    Root world link-->
    <link name="world"/>


    <!-- Obstacle link -->
    <link name="obstacle_link">
        <inertial>
                <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
                <mass value="0.01"/>
                <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
            </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
               <sphere radius="0.4"/>
            </geometry>
            <material name="Cyan">
               <color rgba="0 1.0 1.0 0.1"/>
            </material>
        </visual>
    </link>
    <joint name="obstacle_joint" type="fixed">
      <parent link="omni_base_root_link"/>
      <child link="obstacle_link"/>
      <origin xyz="0.0 0.0 1.2" rpy="0 0 0"/>
    </joint>

    <!-- Base limits link -->
    <link name="base_limit_left">
        <visual>
            <origin xyz="0 ${0.4 + 0.5/2 + 0.35} 0" rpy="0 0 0" /> <!-- width_robot/2 + wight_box_marker/2 + clearance -->
            <geometry>
               <box size="20 0.5 0.2"/>
            </geometry>
            <material name="Red">
               <color rgba="1.0 0.0 0.0 0.5"/>
            </material>
        </visual>
    </link>
    <joint name="base_limit_left_joint" type="fixed">
      <parent link="omni_base_root_link"/>
      <child link="base_limit_left"/>
      <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
    </joint>

    <link name="base_limit_right">
        <visual>
            <origin xyz="0 -${0.4 + 0.5/2 + 0.35} 0" rpy="0 0 0" /> <!-- width_robot/2 + wight_box_marker/2 + clearance -->
            <geometry>
               <box size="20 0.5 0.2"/>
            </geometry>
            <material name="Red">
               <color rgba="1.0 0.0 0.0 0.5"/>
            </material>
        </visual>
    </link>
    <joint name="base_limit_right_joint" type="fixed">
      <parent link="omni_base_root_link"/>
      <child link="base_limit_right"/>
      <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
    </joint>
    <!--    Omnidirectional base-->
    <xacro:unless value="$(arg fixed_base)">
        <joint name="world_joint" type="fixed">
            <parent link="world"/>
            <child link="omni_base_root_link"/>
        </joint>
        <xacro:planar_base connect_to="base_link"/>
    </xacro:unless>

    <xacro:if value="$(arg fixed_base)">
        <joint name="world_joint" type="fixed">
            <parent link="world"/>
            <child link="base_link"/>
             <origin xyz="-0.5265965085817698 -0.2806388380571893 0.0" rpy="0.000 0.000 -0.04332860681335808"/>
        </joint>
    </xacro:if>

    <!--   Ridgeback base: for collision and appearance -->
    <xacro:ridgeback/>

    <!--Mount panda on ridgeback-->
    <joint name="arm_mount_joint" type="fixed">
        <parent link="mid_mount"/>
        <child link="franka_mount_link"/>
        <!--        <origin xyz="$(optenv RIDGEBACK_FRANKA_XYZ 0.21 0.0 0.2)" rpy="$(optenv RIDGEBACK_FRANKA_RPY 0 0 0)"/>-->
        <origin xyz="0.282 0.005 0.447" rpy="0 0 0"/>

    </joint>
    <link name="franka_mount_link"/>
    <xacro:panda_robot parent_link="franka_mount_link"/>

    <!--    Add utility grasp frame-->
    <xacro:if value="${add_finger}">
        <link name="panda_grasp"/>
        <joint name="panda_grasp_joint" type="fixed">
            <parent link="panda_leftfinger"/>
            <child link="panda_grasp"/>
            <axis xyz="0 1 0"/>
            <origin xyz="0 -0.02 0.035" rpy="0 0 0"/> <!-- 0 0 0.45 offset panda_hand xyz="0 0 0.095" rpy="0 0 0"-->
        </joint>
    </xacro:if>

    <!-- The position of the frame published by Vicon -->
    <link name="reference_link"/>
    <joint name="reference_link_joint" type="fixed">
        <origin xyz="-0.015 0.0 0.01" rpy="0 0 0"/>
        <parent link="franka_mount_link"/>
        <child link="reference_link"/>
    </joint>


</robot>